// @microsoft/teamsfx-core/build/component/feature/cicd/questions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.questionNames = exports.providerIdToLabel = exports.templateIdToLabel = exports.templateOptions = exports.publishOption = exports.provisionOption = exports.cdOption = exports.ciOption = exports.providerOptions = exports.jenkinsOption = exports.azdoOption = exports.githubOption = void 0;
const localizeUtils_1 = require("../../../common/localizeUtils");
exports.githubOption = {
    id: "github",
    label: "GitHub",
    detail: "",
};
exports.azdoOption = {
    id: "azdo",
    label: "Azure DevOps",
    detail: "",
};
exports.jenkinsOption = {
    id: "jenkins",
    label: "Jenkins",
    detail: "",
};
exports.providerOptions = [exports.githubOption, exports.azdoOption, exports.jenkinsOption];
exports.ciOption = {
    id: "ci",
    label: "CI",
    detail: localizeUtils_1.getLocalizedString("plugins.cicd.ciOption.detail"),
};
exports.cdOption = {
    id: "cd",
    label: "CD",
    detail: localizeUtils_1.getLocalizedString("plugins.cicd.cdOption.detail"),
};
exports.provisionOption = {
    id: "provision",
    label: "Provision",
    detail: localizeUtils_1.getLocalizedString("plugins.cicd.provisionOption.detail"),
};
exports.publishOption = {
    id: "publish",
    label: "Publish to Teams",
    detail: localizeUtils_1.getLocalizedString("plugins.cicd.publishOption.detail"),
};
exports.templateOptions = [exports.ciOption, exports.provisionOption, exports.cdOption, exports.publishOption];
const templateIdLabelMap = new Map([
    [exports.ciOption.id, exports.ciOption.label],
    [exports.cdOption.id, exports.cdOption.label],
    [exports.provisionOption.id, exports.provisionOption.label],
    [exports.publishOption.id, exports.publishOption.label],
]);
const providerIdLabelMap = new Map([
    [exports.githubOption.id, exports.githubOption.label],
    [exports.azdoOption.id, exports.azdoOption.label],
    [exports.jenkinsOption.id, exports.jenkinsOption.label],
]);
function templateIdToLabel(templateId) {
    var _a;
    return (_a = templateIdLabelMap.get(templateId)) !== null && _a !== void 0 ? _a : templateId;
}
exports.templateIdToLabel = templateIdToLabel;
function providerIdToLabel(providerId) {
    var _a;
    return (_a = providerIdLabelMap.get(providerId)) !== null && _a !== void 0 ? _a : providerId;
}
exports.providerIdToLabel = providerIdToLabel;
var questionNames;
(function (questionNames) {
    questionNames["Provider"] = "provider";
    questionNames["Template"] = "template";
    questionNames["Environment"] = "target-env";
})(questionNames = exports.questionNames || (exports.questionNames = {}));
