// @microsoft/teamsfx-core/build/component/driver/deploy/azure/azureAppServiceDeployDriver.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureAppServiceDeployDriverImpl = exports.AzureAppServiceDeployDriver = void 0;
const tslib_1 = require("tslib");
const azureDeployDriver_1 = require("./azureDeployDriver");
const typedi_1 = require("typedi");
const common_1 = require("../../../utils/common");
const lib_1 = require("@feathersjs/hooks/lib");
const addStartAndEndTelemetry_1 = require("../../middleware/addStartAndEndTelemetry");
const commonConstant_1 = require("../../../constant/commonConstant");
const deployConstant_1 = require("../../../constant/deployConstant");
const messages_1 = require("../../../messages");
const localizeUtils_1 = require("../../../../common/localizeUtils");
const ACTION_NAME = "azureAppService/deploy";
let AzureAppServiceDeployDriver = class AzureAppServiceDeployDriver {
    constructor() {
        this.description = localizeUtils_1.getLocalizedString("driver.deploy.deployToAzureAppServiceDescription");
    }
    async run(args, context) {
        const impl = new AzureAppServiceDeployDriverImpl(args, context);
        return common_1.wrapRun(() => impl.run(), () => impl.cleanup(), context.logProvider);
    }
    execute(args, ctx) {
        return common_1.wrapSummary(this.run.bind(this, args, ctx), [
            "driver.deploy.azureAppServiceDeploySummary",
        ]);
    }
};
tslib_1.__decorate([
    lib_1.hooks([addStartAndEndTelemetry_1.addStartAndEndTelemetry(ACTION_NAME, commonConstant_1.TelemetryConstant.DEPLOY_COMPONENT_NAME)]),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object, Object]),
    tslib_1.__metadata("design:returntype", Promise)
], AzureAppServiceDeployDriver.prototype, "run", null);
AzureAppServiceDeployDriver = tslib_1.__decorate([
    typedi_1.Service(ACTION_NAME)
], AzureAppServiceDeployDriver);
exports.AzureAppServiceDeployDriver = AzureAppServiceDeployDriver;
class AzureAppServiceDeployDriverImpl extends azureDeployDriver_1.AzureDeployDriver {
    constructor() {
        super(...arguments);
        this.pattern = /\/subscriptions\/([^\/]*)\/resourceGroups\/([^\/]*)\/providers\/Microsoft.Web\/sites\/([^\/]*)/i;
        this.helpLink = "https://aka.ms/teamsfx-actions/azure-app-service-deploy";
    }
    async azureDeploy(args, azureResource, azureCredential) {
        var _a, _b, _c, _d;
        await ((_a = this.progressBar) === null || _a === void 0 ? void 0 : _a.start());
        const cost = await this.zipDeploy(args, azureResource, azureCredential);
        await ((_b = this.progressBar) === null || _b === void 0 ? void 0 : _b.next(messages_1.ProgressMessages.restartAzureService));
        await this.restartFunctionApp(azureResource);
        await ((_c = this.progressBar) === null || _c === void 0 ? void 0 : _c.end(true));
        if (cost > deployConstant_1.DeployConstant.DEPLOY_OVER_TIME) {
            await ((_d = this.context.logProvider) === null || _d === void 0 ? void 0 : _d.info(messages_1.getLocalizedMessage("driver.deploy.notice.deployAcceleration", "https://learn.microsoft.com/en-us/azure/app-service/deploy-run-package").localized));
        }
    }
    createProgressBar(ui) {
        var _a;
        return ui === null || ui === void 0 ? void 0 : ui.createProgressBar(`Deploying ${(_a = this.workingDirectory) !== null && _a !== void 0 ? _a : ""} to Azure App Service`, 6);
    }
}
exports.AzureAppServiceDeployDriverImpl = AzureAppServiceDeployDriverImpl;
