// @microsoft/teamsfx-core/build/core/middleware/utils/debug/appLocalYmlGenerator.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppLocalYmlGenerator = exports.AppLocalYmlConfig = void 0;
const appYmlGenerator_1 = require("../appYmlGenerator");
class AppLocalYmlConfig {
}
exports.AppLocalYmlConfig = AppLocalYmlConfig;
class AppLocalYmlGenerator extends appYmlGenerator_1.BaseAppYmlGenerator {
    constructor(oldProjectSettings, config, placeholderMappings) {
        super(oldProjectSettings);
        this.handlebarsContext = {
            config: config,
            placeholderMappings: placeholderMappings,
        };
    }
    async generateAppYml() {
        var _a;
        this.generateHandlerbarsContext();
        switch ((_a = this.oldProjectSettings.programmingLanguage) === null || _a === void 0 ? void 0 : _a.toLowerCase()) {
            case "javascript":
            case "typescript":
            default:
                // only support js/ts at first
                return await this.buildHandlebarsTemplate("js.ts.app.local.yml");
        }
    }
    async generateHandlerbarsContext() {
        var _a;
        const azureSolutionSettings = this.oldProjectSettings.solutionSettings;
        let functionName = undefined;
        if (azureSolutionSettings.activeResourcePlugins.includes("fx-resource-function")) {
            functionName = this.oldProjectSettings.defaultFunctionName || "getUserProfile";
        }
        if ((_a = this.handlebarsContext.config.deploy) === null || _a === void 0 ? void 0 : _a.sso) {
            if (azureSolutionSettings.activeResourcePlugins.includes("fx-resource-frontend-hosting")) {
                this.handlebarsContext.config.deploy.ssoTab = {
                    functionName,
                };
            }
            if (azureSolutionSettings.activeResourcePlugins.includes("fx-resource-bot")) {
                this.handlebarsContext.config.deploy.ssoBot = true;
            }
            if (functionName) {
                this.handlebarsContext.config.deploy.ssoFunction = true;
            }
        }
    }
}
exports.AppLocalYmlGenerator = AppLocalYmlGenerator;
