// @microsoft/teamsfx-core/build/common/versionMetadata.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.VersionState = exports.Metadata = exports.MetadataV2 = exports.MetadataV3 = void 0;
exports.MetadataV3 = {
    vscodeStarterVersion: "5.0.0",
    // TODO: it should be confirmed
    cliStarterVersion: "2.0.0",
    // TODO: it should be confirmed
    vsStarterVersion: "5.0.0",
    projectVersion: "3.0.0",
};
exports.MetadataV2 = {
    projectVersion: "2.0.0",
    projectMaxVersion: "2.1.0",
};
exports.Metadata = {
    versionMatchLink: "https://aka.ms/teamsfx-project-toolkit-match",
};
var VersionState;
(function (VersionState) {
    // project version compatible
    VersionState[VersionState["compatible"] = 0] = "compatible";
    // project version outdated, project should upgrade
    VersionState[VersionState["upgradeable"] = -1] = "upgradeable";
    // project version ahead, need update toolkit
    VersionState[VersionState["unsupported"] = 1] = "unsupported";
})(VersionState = exports.VersionState || (exports.VersionState = {}));
